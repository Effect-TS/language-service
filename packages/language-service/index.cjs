// eslint-disable-next-line @typescript-eslint/consistent-type-imports, no-undef
module.exports = function init(modules) {
  function create(info) {
    // start with regular language service, then load custom once async module has loaded
    let languageService = info.languageService
    import("./build/cjs/index.js").then(v => languageService = v.default(modules).create(info))

    // create the proxy
    const proxy = Object.create(null)
    for (const k of Object.keys(info.languageService)) {
      proxy[k] = (...args) => languageService[k].apply(languageService, args)
    }

    return proxy
  }

  return { create }
}
