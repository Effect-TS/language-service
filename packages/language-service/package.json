{
  "name": "@effect/language-service",
  "version": "0.0.0",
  "type": "module",
  "repository": "https://github.com/Effect-TS/language-service.git",
  "author": "Mattia Manzati",
  "license": "MIT",
  "publishConfig": {
    "access": "public",
    "directory": "dist"
  },
  "scripts": {
    "clean": "yarn org:rimraf build tsbuildinfo dist",
    "lint": "yarn org:eslint . --ext .ts,.tsx",
    "autofix": "yarn lint --fix"
  },
  "main": "./index.cjs",
  "exports": {
    ".": {
      "import": "./build/cjs/index.js"
    },
    "./*": {
      "import": "./build/cjs/*.js"
    },
    "./test/*": {
      "import": "./build/test/*.js"
    }
  },
  "dependencies": {
    "@effect/core": "^0.0.15",
    "@tsplus/runtime": "^0.1.10",
    "@tsplus/stdlib": "^0.2.8"
  },
  "config": {
    "side": [],
    "modules": []
  }
}
