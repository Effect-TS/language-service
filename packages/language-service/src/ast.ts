import { Tag } from "@tsplus/stdlib/service/Tag"
import type ts from "typescript/lib/tsserverlibrary"

declare module "typescript/lib/tsserverlibrary" {
  const nullTransformationContext: ts.TransformationContext

  export namespace formatting {
    interface FormattingHost {
      getNewLine?(): string
    }

    export interface FormatContext {
      readonly options: ts.FormatCodeSettings
      readonly getRules: unknown
    }

    function getFormatContext(options: ts.FormatCodeSettings, host: FormattingHost): FormatContext
  }

  export type TextChangesContext = any
  export type ChangeNodeOptions = any

  export namespace textChanges {
    export class ChangeTracker {
      static with(context: ts.TextChangesContext, cb: (tracker: ChangeTracker) => void): ts.FileTextChanges[]
      replaceNode(sourceFile: ts.SourceFile, oldNode: ts.Node, newNode: ts.Node, options?: ts.ChangeNodeOptions): void
    }
    export function applyChanges(text: string, changes: readonly ts.TextChange[]): string
  }

  export function findPrecedingToken(
    position: number,
    sourceFile: ts.SourceFileLike,
    startNode: ts.Node,
    excludeJsdoc?: boolean
  ): ts.Node | undefined
  export function findPrecedingToken(
    position: number,
    sourceFile: ts.SourceFile,
    startNode?: ts.Node,
    excludeJsdoc?: boolean
  ): ts.Node | undefined

  export function isMemberName(node: ts.Node): node is ts.MemberName
  export function isKeyword(token: ts.SyntaxKind): token is ts.KeywordSyntaxKind
}

// eslint-disable-next-line @typescript-eslint/consistent-type-imports
export type TypeScriptApi = typeof import("typescript/lib/tsserverlibrary")
export const TypeScriptApi = Tag<TypeScriptApi>()
export const LanguageServiceApi = Tag<ts.LanguageService>()
export const ChangeTrackerApi = Tag<ts.textChanges.ChangeTracker>()

export class NoTypeScriptProgramError {
  readonly _tag = "NoTypeScriptProgramError"
}

export function getProgram() {
  return Effect.serviceWithEffect(
    LanguageServiceApi,
    languageService => Effect.sync(() => languageService.getProgram())
  )
    .filterOrFail((program): program is ts.Program => !!program, () => new NoTypeScriptProgramError())
}

export class NoSuchSourceFile {
  readonly _tag = "NoSuchSourceFile"
  constructor(
    readonly fileName: string
  ) {}
}

export function getSourceFile(fileName: string) {
  return getProgram().map(program => program.getSourceFile(fileName))
    .filterOrFail((sourceFile): sourceFile is ts.SourceFile => !!sourceFile, () => new NoSuchSourceFile(fileName))
}

export function hasModifier(node: ts.Declaration, kind: ts.ModifierFlags) {
  return Effect.serviceWith(TypeScriptApi, ts => !!(ts.getCombinedModifierFlags(node) & kind))
}

function getChildNodesContainingRange(
  parent: ts.Node,
  textRange: ts.TextRange
) {
  return Do($ => {
    const ts = $(Effect.service(TypeScriptApi))
    let result: Chunk<ts.Node> = Chunk.empty()

    ts.forEachChild(parent, node => {
      if (node.end < textRange.pos) {
        return
      } else if (node.pos > textRange.end) {
        return
      } else {
        result = result.append(node)
      }
    })

    return result
  })
}

/**
 * Gets the closest node that contains given TextRange
 */
export function getNodesContainingRange(
  sourceFile: ts.SourceFile,
  textRange: ts.TextRange
) {
  return Do(($) => {
    let result: Chunk<ts.Node> = Chunk.empty()

    $(
      Effect.iterate([sourceFile] as ts.Node[], (_) => _.length > 0)(toProcess => {
        const node = toProcess.shift()!
        return getChildNodesContainingRange(node, textRange).flatMap(_ =>
          Effect.sync(() => {
            result = result.concat(_)
            return toProcess.concat(..._.toArray)
          })
        )
      })
    )

    return Chunk.from(result.reverse)
  })
}

/**
 * Ensures value is a text range
 */
export function toTextRange(positionOrRange: number | ts.TextRange): ts.TextRange {
  return typeof positionOrRange === "number" ? { end: positionOrRange, pos: positionOrRange } : positionOrRange
}

export function getHumanReadableName(sourceFile: ts.SourceFile, node: ts.Node) {
  const text = node.getText(sourceFile)
  return text.length > 10 ? text.substring(0, 10) + "..." : text
}

export function collectAll<R, E>(rootNode: ts.Node, test: (node: ts.Node) => Effect<R, E, boolean>) {
  return Do(($) => {
    const ts = $(Effect.service(TypeScriptApi))
    const env = $(Effect.environment<R>())
    let result = Chunk.empty<ts.Node>()

    function visitor(node: ts.Node) {
      if (test(node).provideEnvironment(env).unsafeRunSync()) result = result.append(node)
      ts.visitEachChild(node, visitor, ts.nullTransformationContext)

      return node
    }

    ts.visitNode(rootNode, visitor)

    return result
  })
}

export function getRelevantTokens(
  position: number,
  sourceFile: ts.SourceFile
) {
  return Do(($) => {
    const ts = $(Effect.service(TypeScriptApi))
    const previousToken = ts.findPrecedingToken(position, sourceFile)
    if (
      previousToken && position <= previousToken.end &&
      (ts.isMemberName(previousToken) || ts.isKeyword(previousToken.kind))
    ) {
      const contextToken = ts.findPrecedingToken(previousToken.getFullStart(), sourceFile, /*startNode*/ undefined)! // TODO: GH#18217
      return { contextToken, previousToken }
    }
    return { contextToken: Maybe.fromNullable(previousToken), previousToken: Maybe.fromNullable(previousToken) }
  })
}
