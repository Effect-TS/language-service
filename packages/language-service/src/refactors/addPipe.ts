import * as T from "@effect/core/io/Effect"
import * as AST from "@effect/language-service/ast"
import { createRefactor } from "@effect/language-service/refactors/definition"
import { asPipeArguments, isPipeableCallExpression } from "@effect/language-service/utils"

export default createRefactor({
  name: "effect/addPipe",
  description: "Rewrite using pipe",
  apply: (sourceFile, textRange) =>
    Do($ => {
      const ts = $(T.service(AST.TypeScriptApi))

      const nodes = $(AST.getNodesContainingRange(sourceFile, textRange))
      const pipeableCallExpressions = $(Effect.filterPar(nodes.reverse, isPipeableCallExpression))

      return pipeableCallExpressions.filter(ts.isCallExpression).head.map(node => ({
        description: `Rewrite ${AST.getHumanReadableName(sourceFile, node.expression)} to pipe`,
        apply: Do($ => {
          const changeTracker = $(T.service(AST.ChangeTrackerApi))
          const args = $(asPipeArguments(node))

          const newNode = ts.factory.createCallExpression(
            ts.factory.createIdentifier("pipe"),
            undefined,
            args.toArray
          )

          changeTracker.replaceNode(sourceFile, node, newNode)
        })
      }))
    })
})
