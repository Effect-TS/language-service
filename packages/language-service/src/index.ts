import * as T from "@effect/core/io/Effect"
import * as AST from "@effect/language-service/ast"
import type { DiagnosticDefinition } from "@effect/language-service/diagnostics/definition"
import diagnostics from "@effect/language-service/diagnostics/index"
import type { RefactorDefinition } from "@effect/language-service/refactors/definition"
import refactors from "@effect/language-service/refactors/index"
import { identity } from "@tsplus/stdlib/data/Function"
import * as O from "@tsplus/stdlib/data/Maybe"

// eslint-disable-next-line @typescript-eslint/consistent-type-imports
export default function init(modules: { typescript: typeof import("typescript/lib/tsserverlibrary") }) {
  const ts = modules.typescript

  function create(info: ts.server.PluginCreateInfo) {
    const languageService = info.languageService

    // create the proxy
    const proxy: ts.LanguageService = Object.create(null)
    for (const k of Object.keys(info.languageService) as Array<keyof ts.LanguageService>) {
      // @ts-expect-error
      proxy[k] = (...args: Array<{}>) => languageService[k]!.apply(languageService, args)
    }

    proxy.getSuggestionDiagnostics = (...args) => {
      const suggestionDiagnostics = languageService.getSuggestionDiagnostics(...args)
      const [fileName] = args

      const effectDiagnostics = AST.getSourceFile(fileName).flatMap(sourceFile =>
        T.forEachPar(
          Object.values<DiagnosticDefinition>(diagnostics),
          diagnostic =>
            diagnostic.apply(sourceFile).map(results =>
              results.map(_ => ({
                file: sourceFile,
                start: _.node.pos,
                length: _.node.end - _.node.pos,
                messageText: _.messageText,
                category: _.category,
                code: diagnostic.code
              }))
            )
        )
      ).map(v => v.flatten).map(v => Array.from(v))
        .provideService(AST.LanguageServiceApi, info.languageService)
        .provideService(AST.TypeScriptApi, modules.typescript)
        .unsafeRunSync()

      return suggestionDiagnostics.concat(effectDiagnostics)
    }

    proxy.getApplicableRefactors = (...args) => {
      const applicableRefactors = languageService.getApplicableRefactors(...args)
      const [fileName, positionOrRange] = args
      const textRange = AST.toTextRange(positionOrRange)

      const effectRefactors = AST.getSourceFile(fileName).flatMap(sourceFile =>
        T.collectAllWith(
          Object.values<RefactorDefinition>(refactors).map(refactor =>
            refactor.apply(sourceFile, textRange).map(canApply =>
              canApply.map(_ => ({
                name: refactor.name,
                description: refactor.description,
                actions: [{
                  name: refactor.name,
                  description: _.description
                }]
              }))
            )
          ),
          identity
        )
      ).map(v => Array.from(v))
        .provideService(AST.LanguageServiceApi, info.languageService)
        .provideService(AST.TypeScriptApi, modules.typescript)
        .unsafeRunSync()

      info.project.projectService.logger.info(
        "[@effect/language-service] possible refactors are " + JSON.stringify(effectRefactors)
      )

      return applicableRefactors.concat(effectRefactors)
    }

    proxy.getEditsForRefactor = (
      fileName,
      formatOptions,
      positionOrRange,
      refactorName,
      actionName,
      preferences,
      ...args
    ) => {
      for (const refactor of Object.values(refactors)) {
        if (refactor.name === refactorName) {
          return Do($ => {
            const sourceFile = $(AST.getSourceFile(fileName))
            const textRange = AST.toTextRange(positionOrRange)
            const possibleRefactor = $(refactor.apply(sourceFile, textRange))

            if (O.isNone(possibleRefactor)) {
              info.project.projectService.logger.info(
                "[@effect/language-service] requested refactor " + refactorName + " is not applicable"
              )
              return { edits: [] }
            }

            const formatContext = ts.formatting.getFormatContext(formatOptions, info.languageServiceHost)
            const edits = ts.textChanges.ChangeTracker.with(
              {
                formatContext,
                host: info.languageServiceHost,
                preferences: preferences || {}
              },
              changeTracker =>
                possibleRefactor.value.apply.provideService(AST.ChangeTrackerApi, changeTracker).provideService(
                  AST.TypeScriptApi,
                  modules.typescript
                )
                  .provideService(AST.LanguageServiceApi, info.languageService).unsafeRunSync()
            )

            return { edits }
          }).provideService(AST.TypeScriptApi, modules.typescript)
            .provideService(AST.LanguageServiceApi, info.languageService)
            .unsafeRunSync()
        }
      }

      return languageService.getEditsForRefactor(
        fileName,
        formatOptions,
        positionOrRange,
        refactorName,
        actionName,
        preferences,
        ...args
      )
    }

    return proxy
  }

  return { create }
}
