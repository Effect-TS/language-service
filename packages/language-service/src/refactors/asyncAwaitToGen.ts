import * as T from "@effect/core/io/Effect"
import * as AST from "@effect/language-service/ast"
import { createRefactor } from "@effect/language-service/refactors/definition"
import { transformAsyncAwaitToEffectGen } from "@effect/language-service/utils"

export default createRefactor({
  name: "effect/asyncAwaitToGen",
  description: "Convert to Effect.gen",
  apply: (sourceFile, textRange) =>
    Do($ => {
      const ts = $(T.service(AST.TypeScriptApi))

      const nodes = $(AST.getNodesContainingRange(sourceFile, textRange))

      return nodes.filter(ts.isFunctionDeclaration).filter(node => !!node.body).filter(node =>
        !!(ts.getCombinedModifierFlags(node) & ts.ModifierFlags.Async)
      ).head.map(node => ({
        description: "Rewrite to Effect.gen",
        apply: Do($ => {
          const changeTracker = $(T.service(AST.ChangeTrackerApi))

          const newDeclaration = $(transformAsyncAwaitToEffectGen(node, expression =>
            ts.factory.createCallExpression(
              ts.factory.createPropertyAccessExpression(
                ts.factory.createIdentifier("Effect"),
                "promise"
              ),
              undefined,
              [expression]
            )))

          changeTracker.replaceNode(sourceFile, node, newDeclaration)
        })
      }))
    })
})
