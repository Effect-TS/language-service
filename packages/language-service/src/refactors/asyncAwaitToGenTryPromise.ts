import * as T from "@effect/core/io/Effect"
import * as AST from "@effect/language-service/ast"
import { createRefactor } from "@effect/language-service/refactors/definition"
import { transformAsyncAwaitToEffectGen } from "@effect/language-service/utils"

export default createRefactor({
  name: "effect/asyncAwaitToGenTryPromise",
  description: "Convert to Effect.gen with failures",
  apply: (sourceFile, textRange) =>
    Do($ => {
      const ts = $(T.service(AST.TypeScriptApi))

      const nodes = $(AST.getNodesContainingRange(sourceFile, textRange))

      return nodes.filter(ts.isFunctionDeclaration).filter(node => !!node.body).filter(node =>
        !!(ts.getCombinedModifierFlags(node) & ts.ModifierFlags.Async)
      ).head.map(node => ({
        description: "Rewrite to Effect.gen with failures",
        apply: Do($ => {
          const changeTracker = $(T.service(AST.ChangeTrackerApi))
          let errorCount = 0

          function createErrorADT() {
            errorCount++
            return ts.factory.createObjectLiteralExpression([
              ts.factory.createPropertyAssignment(
                "_tag",
                ts.factory.createAsExpression(
                  ts.factory.createStringLiteral("Error" + errorCount),
                  ts.factory.createTypeReferenceNode("const")
                )
              ),
              ts.factory.createShorthandPropertyAssignment("error")
            ])
          }

          const newDeclaration = $(transformAsyncAwaitToEffectGen(node, expression =>
            ts.factory.createCallExpression(
              ts.factory.createPropertyAccessExpression(
                ts.factory.createIdentifier("Effect"),
                "tryCatchPromise"
              ),
              undefined,
              [
                ts.factory.createArrowFunction(undefined, undefined, [], undefined, undefined, expression),
                ts.factory.createArrowFunction(
                  undefined,
                  undefined,
                  [ts.factory.createParameterDeclaration(undefined, undefined, "error")],
                  undefined,
                  undefined,
                  createErrorADT()
                )
              ]
            )))

          changeTracker.replaceNode(sourceFile, node, newDeclaration)
        })
      }))
    })
})
