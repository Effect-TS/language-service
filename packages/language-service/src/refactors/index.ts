import addPipe from "@effect/language-service/refactors/addPipe"
import asyncAwaitToGen from "@effect/language-service/refactors/asyncAwaitToGen"
import asyncAwaitToGenTryPromise from "@effect/language-service/refactors/asyncAwaitToGenTryPromise"
import functionToArrow from "@effect/language-service/refactors/functionToArrow"
import removeCurryArrow from "@effect/language-service/refactors/removeCurryArrow"
import removePipe from "@effect/language-service/refactors/removePipe"

export default { asyncAwaitToGen, asyncAwaitToGenTryPromise, removePipe, addPipe, removeCurryArrow, functionToArrow }
