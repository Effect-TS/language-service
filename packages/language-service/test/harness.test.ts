import * as AST from "@effect/language-service/ast"
import type { RefactorDefinition } from "@effect/language-service/refactors/definition"
import refactors from "@effect/language-service/refactors/index"
import * as O from "@tsplus/stdlib/data/Maybe"
import * as fs from "fs"
import ts from "typescript/lib/tsserverlibrary"

function createMockLanguageServiceHost(fileName: string, sourceText: string): ts.LanguageServiceHost {
  return {
    getCompilationSettings() {
      return {
        ...ts.getDefaultCompilerOptions(),
        strict: true,
        target: ts.ScriptTarget.ESNext,
        noEmit: true,
        moduleResolution: ts.ModuleResolutionKind.NodeJs
      }
    },
    getScriptFileNames() {
      return [fileName]
    },
    getScriptVersion(_fileName) {
      return ""
    },
    getScriptSnapshot(_fileName) {
      if (_fileName === fileName) {
        return ts.ScriptSnapshot.fromString(sourceText)
      }
      return ts.ScriptSnapshot.fromString(fs.readFileSync(_fileName).toString())
    },
    getCurrentDirectory: () => ".",
    getDefaultLibFileName(options) {
      return ts.getDefaultLibFilePath(options)
    },
    fileExists: (_fileName) => {
      if (_fileName === fileName) return true
      return fs.existsSync(_fileName)
    },
    readFile: (_fileName) => {
      if (_fileName === fileName) return sourceText
      return fs.readFileSync(_fileName).toString()
    }
  }
}

/**
 * Loop through text changes, and update start and end positions while running
 */
function forEachTextChange(
  changes: readonly ts.TextChange[],
  cb: (change: ts.TextChange) => void
): void {
  changes = JSON.parse(JSON.stringify(changes))
  for (let i = 0; i < changes.length; i++) {
    const change = changes[i]!
    cb(change)
    const changeDelta = change.newText.length - change.span.length
    for (let j = i + 1; j < changes.length; j++) {
      if (changes[j]!.span.start >= change.span.start) {
        changes[j]!.span.start += changeDelta
      }
    }
  }
}

function applyEdits(edits: readonly ts.FileTextChanges[], fileName: string, sourceText: string): string {
  for (const fileTextChange of edits) {
    if (fileTextChange.fileName === fileName) {
      forEachTextChange(fileTextChange.textChanges, (edit) => {
        const content = sourceText
        const prefix = content.substring(0, edit.span.start)
        const middle = edit.newText
        const suffix = content.substring(edit.span.start + edit.span.length)
        sourceText = prefix + middle + suffix
      })
    }
  }
  return sourceText
}

export function testRefactorOnExample(refactor: RefactorDefinition, fileName: string) {
  const sourceWithMarker = fs.readFileSync(require.resolve(__dirname + "/../../examples/" + fileName))
    .toString("utf8")
  const firstLine = (sourceWithMarker.split("\n")[0] || "").trim()
  for (const [lineAndCol] of firstLine.matchAll(/([0-9]+:[0-9]+)/gm)) {
    // create the language service
    const languageServiceHost = createMockLanguageServiceHost(fileName, sourceWithMarker)
    const languageService = ts.createLanguageService(languageServiceHost, undefined, ts.LanguageServiceMode.Semantic)
    const sourceFile = languageService.getProgram()?.getSourceFile(fileName)
    if (!sourceFile) throw new Error("No source file " + fileName + " in VFS")

    // gets the position to test
    const [line, character] = lineAndCol.split(":")
    const cursorPosition = ts.getPositionOfLineAndCharacter(sourceFile, +line! - 1, +character!)
    const textRange = { pos: cursorPosition, end: cursorPosition + 1 }

    // ensure there are no errors in TS file
    const diagnostics = languageService.getCompilerOptionsDiagnostics()
      .concat(languageService.getSyntacticDiagnostics(fileName))
      .concat(languageService.getSemanticDiagnostics(fileName)).map(diagnostic => {
        const message = ts.flattenDiagnosticMessageText(diagnostic.messageText, "\n")
        if (diagnostic.file) {
          const { character, line } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start!)
          return `  Error ${diagnostic.file.fileName} (${line + 1},${character + 1}): ${message}`
        } else {
          return `  Error: ${message}`
        }
      })
    expect(diagnostics).toEqual([])

    // check and assert the refactor is executable
    const canApply = refactor
      .apply(sourceFile, textRange)
      .provideService(AST.TypeScriptApi, ts)
      .provideService(AST.LanguageServiceApi, languageService)
      .unsafeRunSync()

    expect(O.isSome(canApply)).toBe(true)
    if (O.isNone(canApply)) return

    // run the refactor and ensure it matches the snapshot
    const formatContext = ts.formatting.getFormatContext(
      ts.getDefaultFormatCodeSettings("\n"),
      { getNewLine: () => "\n" }
    )
    const edits = ts.textChanges.ChangeTracker.with(
      {
        formatContext,
        host: languageServiceHost,
        preferences: {}
      },
      (changeTracker) =>
        canApply.value.apply
          .provideService(AST.ChangeTrackerApi, changeTracker)
          .provideService(AST.TypeScriptApi, ts)
          .provideService(AST.LanguageServiceApi, languageService)
          .unsafeRunSync()
    )

    expect(applyEdits(edits, fileName, sourceWithMarker)).toMatchSnapshot()
  }
}

function testRefactor(name: string, refactor: RefactorDefinition, fileNames: string[]) {
  for (const fileName of fileNames) {
    describe(fileName, () => {
      it(fileName, () => {
        testRefactorOnExample(refactor, fileName)
      })
    })
  }
}

Object.keys(refactors).map(refactorName => testRefactor(refactorName, refactors[refactorName]!, [refactorName + ".ts"]))
