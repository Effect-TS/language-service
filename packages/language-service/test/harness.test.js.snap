// Vitest Snapshot v1

exports[`addPipe.ts > addPipe.ts 1`] = `
"// 5:18
import * as T from \\"@effect/core/io/Effect\\"
import { pipe } from \\"@tsplus/stdlib/data/Function\\"

const test = pipe(T.succeed(1), T.map((a: number) => a * 2))
"
`;

exports[`asyncAwaitToGen.ts > asyncAwaitToGen.ts 1`] = `
"// 4:28
import * as Effect from \\"@effect/core/io/Effect\\"

export function refactorMe(arg: string) {
    return Effect.gen(function*($) {
        return yield* $(Effect.promise(Promise.resolve(1)))
    })
}
"
`;

exports[`asyncAwaitToGenTryPromise.ts > asyncAwaitToGenTryPromise.ts 1`] = `
"// 4:28
import * as Effect from \\"@effect/core/io/Effect\\"

export function refactorMe(arg: string) {
    return Effect.gen(function*($) {
        for (let i = 0; i < 10; i++) {
            yield* $(Effect.tryCatchPromise(() => Promise.resolve(i), error => ({ _tag: \\"Error1\\" as const, error })))
        }
        return yield* $(Effect.tryCatchPromise(() => Promise.resolve(arg), error => ({ _tag: \\"Error2\\" as const, error })))
    })
}
"
`;

exports[`functionToArrow.ts > functionToArrow.ts 1`] = `
"// 5:20,12:20,21:5
/**
 * Docs
 */
export const myTest = (name: string) => name.length

/**
 * Docs
 */
export function myTest2(name: string) {
  if (name === \\"LOL\\") return 42
  return name.length
}

class Sample {
  /**
   * Docs
   */
  test(name: string) {
    return name.length
  }
}
"
`;

exports[`functionToArrow.ts > functionToArrow.ts 2`] = `
"// 5:20,12:20,21:5
/**
 * Docs
 */
export function myTest(name: string) {
  return name.length
}

/**
 * Docs
 */
export const myTest2 = (name: string) => {
    if (name === \\"LOL\\")
        return 42
    return name.length
}

class Sample {
  /**
   * Docs
   */
  test(name: string) {
    return name.length
  }
}
"
`;

exports[`functionToArrow.ts > functionToArrow.ts 3`] = `
"// 5:20,12:20,21:5
/**
 * Docs
 */
export function myTest(name: string) {
  return name.length
}

/**
 * Docs
 */
export function myTest2(name: string) {
  if (name === \\"LOL\\") return 42
  return name.length
}

class Sample {
  /**
   * Docs
   */
  /**
     * Docs
     */
    test = (name: string) => name.length
}
"
`;

exports[`removeCurryArrow.ts > removeCurryArrow.ts 1`] = `
"// 7:12
import * as T from \\"@effect/core/io/Effect\\"
import { pipe } from \\"@tsplus/stdlib/data/Function\\"

const test = pipe(
  T.succeed(\\"Hello\\"),
  T.tap(T.log)
)
"
`;

exports[`removePipe.ts > removePipe.ts 1`] = `
"// 5:16
import * as T from \\"@effect/core/io/Effect\\"
import { pipe } from \\"@tsplus/stdlib/data/Function\\"

const test = T.map(a => a * 1)(T.succeed(1))
"
`;
